### ISSUE TITLE
twisted.test.test_policies calls reactor.iterate()

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10795]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1192|
|Type|defect|
|Created|2005-09-19 16:56:43Z|

Attachments:

* [policies_1192.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0ae/0aec17ce3bff113e1c7c16e9d9dcaf90bf590974/c3d27bc0ee95c33e4934e2d1b42d260ea31d6aa9.diff) (17621 bytes) - added by therve on 2006-10-04 14:02:53Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1192 1192
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests tests
time__1127149003000000 1127149003000000
changetime__1191690978000000 1191690978000000
version__ 
owner__ 
cc__exarkun cc__spiv cc__itamarst cc__jml cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
testReadLimit and testWriteLimit are the classic broken tests, I think.  It's
unfortunate these have been disabled for so long.  In order to get them running
again, I think we need to totally decouple them from the real reactor
implementation: there's no way what they do, which is essentially
timing-oriented, will ever be determistic.

These should get fake transports and a fake clock should be used.  It should be
advanced determistically by the test method and at each stage, assertions made
about the state of the fake transports.

There shouldn't need to be any interaction with the reactor.

The Clock in test_task might be a useful reference, although I think it does not
go far enough: rather than explicit calls to pump(), I think there should be a
Deferred set up by the test method which only fires when the entire process is
done.  This probably bears further discussion, though.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

This one is hard. I attach a first try I made. It modifies some factories in t.p.policies to make their callLater customizable. I also corrected a bug in task.Clock (called was not set).

For test_policies, TimeoutTestCase is mostly OK. ThrottlingTestCase is simplified, maybe too much, but tests pass.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [21002]) Work around the initial patch.

Refs [#10795](https://github.com/twisted/twisted/issues/10795).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

This is ready to review.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_6"></a>|@jerub set owner to @therve|
|-|-|

This was a big job getting all this rejigged, I think some good work has been done here.

Some comments on the code that I'd like to see addressed:
 * TimeoutTester.callLater should be overridden in the subclass, I don't like the ```p.callLater = self.clock.callLater``` pattern used in TestTimeout.
 * TimeoutTestCase._serverSetup says ```proto.callLater = self.clock.callLater``` but that test passes without that line. I suggest it be removed.
 * When you worked on TimeoutTester you modified a bunch of methods, including their names, but they don't have docstrings. Please add simple docstrings to these methods.

Address these things and I'll be happy to approve its merging.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [21330]) Add some docstrings, refactor some tests.

Refs [#10795](https://github.com/twisted/twisted/issues/10795)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @jerub|
|-|-|

I think I've addressed all the points above (in test-policies-1192).

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_9"></a>|@jerub set owner to @therve|
|-|-|

Thankyou for all that! Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set status to `closed`|
|-|-|

(In [21333]) Merge test-policies-1192

Author: therve
Reviewer: jerub
Fixes [#10795](https://github.com/twisted/twisted/issues/10795)

Refactor tests of twisted.protocol.policies for not using reactor.iterate()

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
